### ISSUE TITLE
Integration with https://immutables.github.io/

### ISSUE THREAD

**Issue by anton-tregubov** [yandex-qatools/hamcrest-pojo-matcher-generator#27]
I think that https://immutables.github.io/ is greatest tool for generating PoJo. Your tool can let developers (lazy developers)  focus on effectve test instead of "near support" code.

I look at your sources? and see that you filter 

`private static Stream<Element> asFields(Element element) {
        switch (element.getKind()) {
            case FIELD:
                return Stream.of(element);
            case CLASS:
                return element.getEnclosedElements().stream().flatMap(MatcherFactoryGenerator::asFields);
        }
        return Stream.empty();
    }`
and
`ru.yandex.qatools.processors.matcher.gen.processing.MethodsCollector#asMethodSpec`
can take into accoount method (because immutables require to define only methods). 

I think i can do it too, but later...

**Comment by lanwen** [comment]
Yes, I've been thinking a lot about using methods too, but it contains some corner cases we should usually be prepared. So I've put it to the backlog without a deadline. But maybe it's time to start progress on it :)

So if I will start it, I'll write about it here (but I should find some time only after next weekend)

**Comment by anton-tregubov** [comment]
Start working on feature. [(my-fork)](https://github.com/anton-tregubov/hamcrest-pojo-matcher-generator)
Can you give some ideas for 
> some corner cases
because i change some lines and generator work fine for my cases

**Comment by lanwen** [comment]
Should generate
- methods from the parent (if not overridden, with annotation)

Should NOT generate for
- methods from the parent (if overridden, without annotation)
- static methods
- non-public methods
- when using both on a field with getter and on getter (should choose only one)
- for void methods
- for methods returning the same class (in most cases it's like the fluent setters for some predefined values)
- for methods with args (or we should provide an ability to pass args to it, which can be quite complicated for such small library)

Thanks a lot, for your time!

**Comment by anton-tregubov** [comment]
Если напрягает говорить на родном, скажи, я переведу.
Я понял почему я не видел проблем. Я хочу сделать гораздо проще. Добавить поиск методов которые соответствуют BeanProperty понятию. Тоесть get/isXXX свойяствам и не совпадают с публичными полями. Тоесть задача сильно упроаеться. 

Единственный кейз который я действительно забыл учесть - это наследование, но думаю я сделаю и это

**Comment by lanwen** [comment]
Супер тогда. с такими ограничениями действительно будет проще. Спасибо
